mutt/random.c

```diff
@@ -0,0 +1,130 @@
+/**
+ * @file
+ * Random number/string functions
+ *
+ * @authors
+ * Copyright (C) 2020 Richard Russon <rich@flatcap.org>
+ *
+ * @copyright
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, either version 2 of the License, or (at your option) any later
+ * version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @page random Random number/string functions
+ *
+ * Random number/string functions
+ */
+
+#include "config.h"
+#include <stddef.h>
+#include <errno.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include "exit.h"
+#include "logging.h"
+#include "message.h"
+#ifdef HAVE_SYS_RANDOM_H
+#include "sys/random.h"
+#endif
+
+static FILE *fp_random;
+
+static const unsigned char base32[] = "abcdefghijklmnopqrstuvwxyz234567";
+
+/**
+ * mutt_randbuf - Fill a buffer with randomness
+ * @param buf    Buffer for result
+ * @param buflen Size of buffer
+ * @retval  0 Success
+ * @retval -1 Error
+ */
+int mutt_randbuf(void *buf, size_t buflen)
+{
+  if (buflen > 1048576)
+  {
+    mutt_error(_("mutt_randbuf buflen=%zu"), buflen);
+    return -1;
+  }
+
+#ifdef HAVE_GETRANDOM
+  long ret;
+  do
+  {
+    ret = getrandom(buf, buflen, 0);
+  } while ((ret == -1) && (errno == EINTR));
```

Comment:
I think this should loop until `buflen` bytes have been read: `getrandom` might return a short count.

Reply:
done